test/core/load_balancing/bm_picker.cc

```diff
@@ -0,0 +1,260 @@
+    auto parsed_json = JsonParse(std::string(config_json_));
+    CHECK_OK(parsed_json);
+    auto config_parsed =
+        CoreConfiguration::Get().lb_policy_registry().ParseLoadBalancingConfig(
+            *parsed_json);
+    CHECK_OK(config_parsed);
+    config_ = std::move(*config_parsed);
+  }
+
+  RefCountedPtr<LoadBalancingPolicy::SubchannelPicker> GetPicker() {
+    MutexLock lock(&mu_);
+    while (picker_ == nullptr) {
+      cv_.Wait(&mu_);
+    }
+    return picker_;
+  }
+
+  void UpdateState(size_t num_endpoints) {
+    {
+      MutexLock lock(&mu_);
+      picker_ = nullptr;
+      work_serializer_->Schedule(
+          [this, num_endpoints]() {
+            EndpointAddressesList addresses;
+            for (size_t i = 0; i < num_endpoints; i++) {
+              grpc_resolved_address addr;
+              int port = i % 65536;
+              int ip = i / 65536;
+              CHECK(ip < 256);
+              CHECK(grpc_parse_uri(
+                  URI::Parse(absl::StrCat("ipv4:127.0.0.", ip, ":", port))
+                      .value(),
+                  &addr));
+              addresses.emplace_back(addr, ChannelArgs());
+            }
+            CHECK_OK(lb_policy_->UpdateLocked(LoadBalancingPolicy::UpdateArgs{
+                std::make_shared<EndpointAddressesListIterator>(
+                    std::move(addresses)),
+                config_, "", ChannelArgs()}));
+          },
+          DEBUG_LOCATION);
+    }
+    work_serializer_->DrainQueue();
+  }
+
+ private:
+  class SubchannelFake final : public SubchannelInterface {
+   public:
+    explicit SubchannelFake(BenchmarkHelper* helper) : helper_(helper) {}
+
+    void WatchConnectivityState(
+        std::unique_ptr<ConnectivityStateWatcherInterface> unique_watcher)
+        override {
+      AddConnectivityWatcherInternal(
+          std::shared_ptr<ConnectivityStateWatcherInterface>(
+              std::move(unique_watcher)));
+    }
+
+    void CancelConnectivityStateWatch(
+        ConnectivityStateWatcherInterface* watcher) override {
+      MutexLock lock(&helper_->mu_);
+      helper_->connectivity_watchers_.erase(watcher);
+    }
+
+    void RequestConnection() override { LOG(FATAL) << "unimplemented"; }
+
+    void ResetBackoff() override { LOG(FATAL) << "unimplemented"; }
+
+    void AddDataWatcher(
+        std::unique_ptr<DataWatcherInterface> watcher) override {
+      auto* watcher_internal =
+          DownCast<InternalSubchannelDataWatcherInterface*>(watcher.get());
+      if (watcher_internal->type() == HealthProducer::Type()) {
+        AddConnectivityWatcherInternal(
+            DownCast<HealthWatcher*>(watcher_internal)->TakeWatcher());
+      } else {
+        LOG(FATAL) << "unimplemented watcher type: "
+                   << watcher_internal->type();
+      }
+    }
+
+    void CancelDataWatcher(DataWatcherInterface* watcher) override {
+      // LOG(FATAL) << "unimplemented";
```

Comment:
Doesn't this need to remove the watcher, in the case of a health watcher?

Reply:
done